### DISCUSSION TITLE
BoxCoxTransformer Implementation

### DISCUSSION THREAD

**Discussion by jrodenbergrheem** [JoaquinAmatRodrigo/skforecast#858] [Ideas]
Is there an integration for Boxcox within Skforecast??? 

I was hoping you all had a trick to integrate sktime's transformers (BoxCox Guerrero). I have tried utilzing it but it doesn't seem compatible. 

I also tried sklearns "PowerTransformer" and it seems to be attempting to call .ravel() on a dataframe causing an error. 

Can post a reproducible example but I am sure this has been done before,

Thanks!

**Comment by JavierEscobarOrtiz** [comment]
Hello @jrodenbergrheem 

Could you please share your example of implementing the `PowerTransformer`?

I was able to use it with the `ForecasterRecursive`.

```python
# Create and fit forecaster
# ==============================================================================
forecaster = ForecasterRecursive(
                 regressor       = LGBMRegressor(random_state=123, verbose=-1),
                 lags            = 15,
                 window_features = RollingFeatures(stats=['mean'], window_sizes=10),
                 transformer_y   = PowerTransformer(method='box-cox'),
             )

forecaster.fit(y=data_train)

# Predict
# ==============================================================================
predictions = forecaster.predict(steps=36)
predictions.head(3)
```

I am using the data from the user guide: https://skforecast.org/latest/user_guides/autoregresive-forecaster

Thanks,

Javi

**Reply by jrodenbergrheem** [reply]
Hi Javi, sure! I am using it with Recursive Multi-Series.... 
<img width="1120" alt="Screenshot 2024-11-25 at 9 31 19 AM" src="https://github.com/user-attachments/assets/9e064031-e453-4bcf-b13f-c22e0bf01610">
<img width="1331" alt="Screenshot 2024-11-25 at 9 32 15 AM" src="https://github.com/user-attachments/assets/ee612cfe-553d-4465-8788-e0db2446933b">

**Reply by jrodenbergrheem** [reply]
^ Also can confirm the same error is thrown when using traditional sk-learn scalers. I am using the Dictionary approach with exog variables, if that could be the thing causing the error. Notably the function transformer approach works just fine !

**Reply by JavierEscobarOrtiz** [reply]
Hi @jrodenbergrheem,

Thank you for the information. Could you please provide a reproducible example? It is really hard to find the problem without seeing the code.

Could you also try running your code with `transformer_exog = None`. Just to discard it is something reagarding the exog transformations.

Thanks a lot!

Javi

**Reply by jrodenbergrheem** [reply]
[Archive.zip](https://github.com/user-attachments/files/17905444/Archive.zip)

Here is the code and notebook, you can go ahead and try utilizing power transformer within transformer series. (It still gives me this erro when specifying no exog transform) 

I am using pandas 2.2.3, Numpy 1.26.4 and sklearn 1.5.2

**Reply by JavierEscobarOrtiz** [reply]
Sorry @jrodenbergrheem, we are having some difficulty running your code. Could you please simplify it a bit by just adding the skforecast part and where the error occurs?

Another option you can try is to run the user guide: [Global Forecasting Models Time Series of Different Lengths and Different Exogenous Variables](https://skforecast.org/0.14.0/user_guides/multi-series-with-different-length-and-different_exog) using the `PowerTransformer` as `transformer_series`. It should be a similar use case of the one that you have. I have tried this and it works fine. 

Javi

**Reply by jrodenbergrheem** [reply]
Hi Javi, appreciate your responsiveness on this. It looks like I found a fix, you must set the transformer output to default using set_config(transform_output = 'default') from sklearn.... 

Here is an example of doing so. It looks like the code is calling "ravel" on a dataframe (so I assume the default transformer output is pandas for skforecast? Just a guess). 



`#create transformer dictionary (random sample of skus to transform) 

from sklearn.preprocessing import PowerTransformer
from sklearn import set_config

set_config(transform_output='default')

pwr_dict = {sku:PowerTransformer(method='yeo-johnson') for sku in df['unique_id'].sample(1000)} # this error also occurred with minmax scaler when I tried it

pwr_dict['_unknown_level'] = None

#define xgb with differencing for non-stationary skus (just to test the performance)
lgb_diff = LGBMRegressor(objective='mae',random_state=42)

lgb_diff = ForecasterRecursiveMultiSeries(regressor=lgb_diff, # tweedie can't train on negative (differenced data)
                                
                                   transformer_series = pwr_dict,

                                   transformer_exog = date_pipeline,

                               differentiation = 1, 
                               
                               lags=[1,2,3,4,52],

                               window_features = [RollingFeatures(stats=all_feats,
                                                                   window_sizes=all_windows)], # add custom feature to list alongwith built in feats 
                    
                    encoding='ordinal', # allow model to use ts_id_columns as features
                    
                    fit_kwargs={'categorical_feature':['Client','Warehouse','Product']})`

**Comment by JavierEscobarOrtiz** [comment]
Solved in #871
